modules/api/pkg/ee/clusterbackup/backupstoragelocation/handler.go

```diff
@@ -0,0 +1,456 @@
+*/
+
+package backupstoragelocation
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"text/template"
+	"time"
+
+	uuid2 "github.com/google/uuid"
+	"github.com/gorilla/mux"
+	velerov1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+
+	apiv2 "k8c.io/dashboard/v2/pkg/api/v2"
+	clusterbackup "k8c.io/dashboard/v2/pkg/ee/clusterbackup/backup"
+	handlercommon "k8c.io/dashboard/v2/pkg/handler/common"
+	"k8c.io/dashboard/v2/pkg/handler/v1/common"
+	"k8c.io/dashboard/v2/pkg/handler/v2/cluster"
+	"k8c.io/dashboard/v2/pkg/provider"
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/log"
+	"k8c.io/kubermatic/v2/pkg/resources"
+	utilerrors "k8c.io/kubermatic/v2/pkg/util/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	supportedVeleroPlugin    = "aws"
+	credentialsSecretKeyName = "cloud-credentials"
+	accessKeyTemplate        = "awsAccessKeyId"
+	secretAccessKeyTemplate  = "awsSecretAccessKey"
+	clusterIdLabelKey        = "cluster-id"
+	bslRegion                = "region"
+	bslKind                  = "BackupStorageLocation"
+	CBSL                     = "cbsl"
+	backupSyncPeriod         = 30 * time.Second
+)
+
+var credentialsTemplate string = `[default]
+aws_access_key_id = {{ .awsAccessKeyId }}
+aws_secret_access_key = {{ .awsSecretAccessKey }}
+`
+
+// createBSLReq defines HTTP request to create BSL in a user cluster
+// swagger:parameters createBackupStorageLocation
+type createBSLReq struct {
+	cluster.GetClusterReq
+
+	// in: body
+	// required: true
+	Body BSLBody
+}
+
+type BSLBody struct {
+	CBSLName string `json:"cbslName"`
+
+	BSLSpec velerov1.BackupStorageLocationSpec `json:"bslSpec"`
+}
+
+func DecodeCreateBSLReq(c context.Context, r *http.Request) (interface{}, error) {
+	var req createBSLReq
+	cr, err := cluster.DecodeGetClusterReq(c, r)
+
+	if err != nil {
+		return nil, err
+	}
+
+	req.GetClusterReq = cr.(cluster.GetClusterReq)
+
+	if err = json.NewDecoder(r.Body).Decode(&req.Body); err != nil {
+		return nil, err
+	}
+	return req, nil
+}
+
+func (r *createBSLReq) validateCreateBSLReq() error {
+	bslSpec := r.Body.BSLSpec
+	if bslSpec.Provider != supportedVeleroPlugin {
+		return fmt.Errorf("velero plugin not supported other than aws")
+	}
+	if r.Body.CBSLName == "" {
+		return fmt.Errorf("a valid CBSL must be provided for the BSL to be created in the cluster for importing backups")
+	}
+	return nil
+}
+
+// checks whether a user is a global admin or a project admin.
+func validateUser(ctx context.Context, userInfoGetter provider.UserInfoGetter, projectId string) error {
+	userInfo, err := userInfoGetter(ctx, projectId)
+	if err != nil {
+		return common.KubernetesErrorToHTTPError(err)
+	}
+	if !userInfo.IsAdmin {
+		return utilerrors.NewNotAuthorized()
+	}
```

Comment:
Please first go through the code and check what `userInfoGetter` is doing, for someone who is only a project admin the `IsAdmin` field will NEVER be true.